File: backend/app/Models/PrintBatch.php
<?php

namespace App\Models;

use App\Models\Concerns\BelongsToTeam;
use FlutterSdk\MagicStarter\Support\ConditionallyUsesUuids;
use Illuminate\Database\Eloquent\Builder;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Support\Facades\DB;

/**
 * A saved set of labels to print together.
 *
 * The migration carries why this exists (items are added over time and printed once, which is what
 * labelling a delivery or relabelling a shelf actually looks like) and what is deliberately absent
 * from it: no render columns, because the preview is keyed on a hash of the template plus its data and
 * exists BEFORE a batch does, while the user is still choosing a template.
 *
 * ### `printed_at` here is derived from the items, never authored
 *
 * A batch is printed when nothing in it is unprinted. Writing this column directly would let it
 * disagree with the rows it summarises, and the resume query reads the ITEMS, so the disagreement
 * would be invisible until a user asked why a finished batch still offered a reprint. `settle()` is the
 * only writer and it reads before it writes.
 */
final class PrintBatch extends Model
{
    use BelongsToTeam;
    use ConditionallyUsesUuids;

    /** @var list<string> */
    protected $fillable = [
        'name',
        'template',
        'fields',
        'created_by',
    ];

    /**
     * @return array<string, string>
     */
    protected function casts(): array
    {
        return [
            'fields' => 'array',
            'printed_at' => 'datetime',
        ];
    }

    /**
     * The lines, in the order they fill a sheet.
     *
     * Ordered by `position` rather than by creation, because `position` is what makes a partially
     * printed batch able to name a range: "reprint 12 to 24" is meaningless if the list can reorder.
     */
    public function items(): HasMany
    {
        return $this->hasMany(PrintBatchItem::class)->orderBy('position');
    }

    /**
     * Batches with something left to print.
     *
     * The resume query, and it asks the ITEMS rather than this row's own `printed_at`: a batch half
     * printed by a jammed printer has a null here and unprinted rows, and a batch printed in two passes
     * has neither.
     */
    public function scopeUnfinished(Builder $query): Builder
    {
        return $query->whereHas('items', function (Builder $items): void {
            $items->whereNull('printed_at');
        });
    }

    /**
     * The lines a print would cover, so the render path and the resource agree on what "pending" means.
     *
     * **On the model rather than as a public static on a controller**, which is where it started: it
     * holds no HTTP, it was called from a second controller, and a static on a controller is the shape
     * that makes that look normal.
     *
     * The eager load includes `serial.product` because a serial line's NAME comes from the serial's
     * product, and loading only `serial` left that as a lazy query per line. Strict mode is off, so it
     * was silent.
     *
     * @return list<PrintBatchItem>
     */
    public function pendingItems(): array
    {
        return $this->items()
            ->whereNull('printed_at')
            ->with(['product', 'serial.product'])
            ->get()
            ->all();
    }

    /**
     * Whether anything in this batch is still waiting.
     */
    public function isUnfinished(): bool
    {
        return $this->items()->whereNull('printed_at')->exists();
    }

    /**
     * The next position a line would take.
     *
     * `unique(print_batch_id, position)` means two concurrent adds can collide, and the collision is a
     * refusal rather than a silent overwrite. That is the right trade for a list a person is building:
     * a retry lands on the next free slot, while a `max + 1` written without the constraint would give
     * two lines one position and make "reprint 12 to 24" ambiguous forever.
     */
    public function nextPosition(): int
    {
        return (int) $this->items()->max('position') + 1;
    }

    /**
     * Records that [$positions] came off a printer, and closes the batch when nothing is left.
     *
     * **Positions rather than ids, because that is what a person reprinting names.** A jammed printer
     * produces "sheets 1 and 2 came out, start again at 25", and the position is the number the sheet
     * itself carries. Ids would make the client hold a mapping it has no reason to have.
     *
     * @param  list<int>  $positions  Empty means every line in the batch.
     * @return int How many lines this marked.
     */
    public function settle(array $positions = []): int
    {
        return DB::transaction(function () use ($positions): int {
            $query = $this->items()->getQuery();

            if ($positions === []) {
                // **Only what has not been printed, and the omission destroyed records of paper.** The
                // render sends the PENDING lines and the client settles with no positions, so an
                // unfiltered mark hit the already-printed rows too: `print_count + 1` for stickers
                // that did not come off a printer, and a fresh `printed_at` over the timestamp of the
                // pass that did. On a feature whose whole point is paper accuracy that is the worst
                // shape a bug can take.
                //
                // A caller naming positions still reaches a printed row on purpose: that is a reprint,
                // and counting it is what `print_count` is for.
                $query->whereNull('printed_at');
            } else {
                $query->whereIn('position', $positions);
            }

            $marked = 0;

            foreach ($query->get() as $item) {
                $item->markPrinted();
                $marked++;
            }

            // Read after the writes, because "printed" is a fact about the items. A batch printed in
            // two passes closes on the second one without anybody having to say so.
            if ($marked > 0 && ! $this->isUnfinished()) {
                $this->forceFill(['printed_at' => $this->freshTimestamp()])->save();
            }

            return $marked;
        });
    }
}

File: lib/app/models/print_batch.dart
import 'package:magic/magic.dart';

import '../../ui/components/label_item_row/label_item_row.dart' show LabelCountMode;

/// One line of a print batch, as the label screen needs it.
///
/// **Replaces `LabelItemFixture`**, which carried the same five fields: `flutter-app.md` says to
/// replace a fixture rather than shadow it, because two types for one thing diverge the moment the API
/// changes.
///
/// `position` is the load-bearing field: it is what `settle`, the copy change and the removal are all
/// keyed on, because it is the number a person reprinting names off a sheet. Ids exist on the wire
/// too, and a client holding one already holds the other.
///
/// **Read-only from this client's side of the write path.** Every mutation goes through
/// `LabelBatchController`'s own endpoints, which answer with the whole batch again; nothing here
/// calls `save()`, so [fillable] stays empty.
class PrintBatchLine extends Model {
  @override
  String get table => 'print_batch_lines';

  @override
  String get resource => 'labels/batches';

  /// Set to false because this app uses string UUIDs as primary keys.
  @override
  bool get incrementing => false;

  @override
  List<String> get fillable => [];

  @override
  Map<String, dynamic> get casts => <String, dynamic>{
    'position': 'int',
    'count': 'int',
    'print_count': 'int',
    'is_printed': 'bool',
  };

  /// The number this line carries on the sheet's list. Unique within a batch, never renumbered.
  int get position => get<int>('position') ?? 0;

  /// The product name, the label's first line.
  String get name => get<String>('name') ?? '';

  /// The code that will be printed.
  ///
  /// **Null means the server did not say, not that one will be generated**, and the difference cost a
  /// whole feature: the resource used to omit this field entirely, so this was always null and every
  /// consumer ran on a placeholder. What actually gets printed is a policy the server owns (a GTIN, a
  /// Code 128 row, or a generated `DPL` code), which is why it travels rather than being guessed.
  String? get code => get<String>('code');

  /// The serial this line prints, when it prints one.
  String? get serial => get<String>('serial');

  /// How many stickers this line contributes.
  int get count => get<int>('count') ?? 1;

  /// Where the count comes from (D45).
  ///
  /// Read straight off the raw string rather than through an [EnumCast]: the wire spells the second
  /// mode `per_serial` and the Dart enum spells it `perSerial`, so a name-matching cast would never
  /// resolve it. The safe fallback is the mode WITHOUT a stepper, matching the server's own default:
  /// offering to edit how many units exist is the mistake D45 exists to prevent.
  LabelCountMode get mode => get<String>('mode') == 'free' ? LabelCountMode.free : LabelCountMode.perSerial;

  /// Whether this line has already been printed in this batch.
  bool get isPrinted => get<bool>('is_printed') ?? false;

  /// How many times it has been printed. Two means two sheets of paper went.
  int get printCount => get<int>('print_count') ?? 0;

  /// Whether this line's copies may be changed.
  ///
  /// D45 twice over: a serial prints once, and a printed line's count is a record of paper that
  /// already went rather than a number to edit.
  bool get isAdjustable => mode == LabelCountMode.free && !isPrinted;

  /// Creates a [PrintBatchLine] for a fixture or a preview, not for hydrating a payload.
  ///
  /// Writes the same wire shape [fromApi] reads, so a getter never has to ask which path built it.
  PrintBatchLine({
    required int position,
    required String name,
    required int count,
    String? code,
    String? serial,
    LabelCountMode mode = LabelCountMode.free,
    bool isPrinted = false,
    int printCount = 0,
  }) {
    setRawAttributes(<String, dynamic>{
      'position': position,
      'name': name,
      'code': code,
      'serial': serial,
      'count': count,
      'mode': mode == LabelCountMode.free ? 'free' : 'per_serial',
      'is_printed': isPrinted,
      'print_count': printCount,
    }, sync: true);
  }

  PrintBatchLine._raw();

  /// The line a batch payload describes.
  static PrintBatchLine fromApi(Map<String, dynamic> json) {
    return PrintBatchLine._raw()
      ..setRawAttributes(json, sync: true)
      ..exists = true;
  }
}

/// A saved set of labels to print together.
///
/// **Read-only from this client's side of the write path.** Every mutation goes through
/// `LabelBatchController`'s own endpoints, which answer with the whole batch again; nothing here
/// calls `save()`, so [fillable] stays empty.
class PrintBatch extends Model {
  @override
  String get table => 'print_batches';

  @override
  String get resource => 'labels/batches';

  /// Set to false because this app uses string UUIDs as primary keys.
  @override
  bool get incrementing => false;

  @override
  List<String> get fillable => [];

  @override
  Map<String, dynamic> get casts => <String, dynamic>{
    'sticker_count': 'int',
    'pending_sticker_count': 'int',
    'printed_at': 'datetime',
  };

  /// The nested line list. `getRelations` resolves it whether the attribute holds raw maps (a
  /// `labels/batches` payload, under its own wire key `items`) or already-built [PrintBatchLine]
  /// instances (a fixture), since it checks `data is List<T>` before reaching for this factory.
  ///
  /// **`PrintBatchLine._raw`, not `PrintBatchLine.new`.** `getRelations` invokes the factory with no
  /// arguments, and the public constructor's `position`/`name`/`count` are required for the fixture
  /// path; the zero-argument private constructor is what a bare hydration needs.
  @override
  Map<String, Model Function()> get relations => <String, Model Function()>{
    'items': PrintBatchLine._raw,
  };

  /// The batch id, or empty before one exists.
  @override
  String get id => get<String>('id') ?? '';

  /// What the user called it, if anything.
  String? get name => get<String>('name');

  /// The sheet template key, for example `a4_8_up_105x70`.
  String get template => get<String>('template') ?? '';

  /// Which fields the label carries.
  ///
  /// A stray non-string entry or a missing/malformed `fields` value both fall back to an EMPTY list
  /// rather than to the constructor's `['name', 'code']` default: an API payload that omitted the
  /// key never meant "the defaults", and guessing otherwise would show fields the server never chose.
  List<String> get fields {
    final Object? raw = getAttribute('fields');

    if (raw is! List) return const <String>[];

    return <String>[for (final Object? field in raw) if (field is String) field];
  }

  /// Every line, in sheet order.
  List<PrintBatchLine> get lines => getRelations<PrintBatchLine>('items');

  /// How many stickers the whole batch is for.
  int get stickerCount => get<int>('sticker_count') ?? 0;

  /// How many stickers a print would produce now.
  int get pendingStickerCount => get<int>('pending_sticker_count') ?? 0;

  /// When the batch finished, or null while anything is left.
  DateTime? get printedAt => get<Carbon>('printed_at')?.toDateTime;

  /// Whether anything is still waiting for a printer.
  ///
  /// Read from the LINES rather than from `printedAt`, which is what the server does: a batch half
  /// printed by a jammed printer has a null date and unprinted rows.
  bool get isUnfinished => lines.any((PrintBatchLine line) => !line.isPrinted);

  /// Whether this is the batch to carry on filling.
  ///
  /// **Not the same question as [isUnfinished], and conflating them stranded batches.** An EMPTY batch
  /// has no unprinted lines, so it read as finished and was never resumed; the client never deletes
  /// one either, so a user who removed their last line got a fresh batch on every visit and the
  /// abandoned ones piled up at the top of a list ordered nulls-first.
  ///
  /// `printedAt` is the server's own answer: it is derived, written only when nothing is left, so a
  /// null means this batch has never been finished. An empty batch is exactly the one to add to.
  bool get isResumable => printedAt == null;

  /// The lines already printed, which stay visible so a batch is resumable rather than mysterious.
  List<PrintBatchLine> get printed =>
      lines.where((PrintBatchLine line) => line.isPrinted).toList(growable: false);

  /// Whether [field] is switched on.
  bool shows(String field) => fields.contains(field);

  /// The codes that will be printed, for the fit check.
  ///
  /// **Only the ones the server named.** This used to substitute `DPL00000000` for a null, on the
  /// premise that null meant "one will be generated"; the resource simply was not sending the field, so
  /// the fit verdict compared a constant against every template's ceiling and was wrong in both
  /// directions. A line whose code is genuinely unknown is left out rather than guessed at, because a
  /// callout naming a code nobody will print is worse than no callout.
  List<String> get codes => <String>[
    for (final PrintBatchLine line in lines)
      if (line.serial != null) line.serial! else if (line.code != null) line.code!,
  ];

  /// Creates a [PrintBatch] for a fixture or a preview, not for hydrating a payload.
  ///
  /// Writes the same wire shape [fromApi] reads (`fields` and `items` mirror the server's own keys),
  /// so a getter never has to ask which path built the instance.
  PrintBatch({
    required String id,
    required String template,
    String? name,
    List<String> fields = const <String>['name', 'code'],
    List<PrintBatchLine> lines = const <PrintBatchLine>[],
    int stickerCount = 0,
    int pendingStickerCount = 0,
    DateTime? printedAt,
  }) {
    setRawAttributes(<String, dynamic>{
      'id': id,
      'name': name,
      'template': template,
      'fields': fields,
      'items': lines,
      'sticker_count': stickerCount,
      'pending_sticker_count': pendingStickerCount,
      'printed_at': printedAt,
    }, sync: true);
  }

  PrintBatch._raw();

  /// The batch a `labels/batches` payload describes.
  static PrintBatch fromApi(Map<String, dynamic> json) {
    return PrintBatch._raw()
      ..setRawAttributes(json, sync: true)
      ..exists = true;
  }
}
